File: algorithms/215. Kth Largest Element in an Array/solution.py
class Solution(object):
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        return Solution.find_kth_num(nums, k, 0, len(nums))

    @staticmethod
    def find_kth_num(nums, k, start, stop):
        if stop - start < 2:
            return nums[start]
        r = nums[random.randrange(start, stop)]
        e = u = start
        g = stop
        while u < g:
            if nums[u] > r:
                nums[u], nums[e] = nums[e], nums[u]
                e += 1
                u += 1
            elif nums[u] < r:
                g -= 1
                nums[u], nums[g] = nums[g], nums[u]
            else:
                u += 1
        if k <= (e-start):
            return Solution.find_kth_num(nums, k, start, e)
        elif k > (g-start):
            return Solution.find_kth_num(nums, k-(g-start), g, stop)
        else:
            return nums[e]

File: algorithms/215. Kth Largest Element in an Array/solution.cc
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        return findKthLargest(nums, k, 0, nums.size());
    }

    int findKthLargest(vector<int> & nums, int k, int start, int end) {
        if ( end - start < 2 ) {
            return nums[start];
        }
        int p = nums[start + rand() % (end - start)];
        int e = start, u = start, g = end;
        while ( u < g ) {
            if ( nums[u] > p ) {
                swap(nums[e], nums[u]);
                ++e;
                ++u;
            } else if ( nums[u] < p ) {
                --g;
                swap(nums[u], nums[g]);
            } else {
                ++u;
            }
        }
        if ( k <= e - start ) {
            return findKthLargest(nums, k, start, e);
        } else if ( k <= g - start ) {
            return nums[e];
        } else {
            return findKthLargest(nums, k - (g - start), g, end);
        }
    }
};
